wagtail/search/query.py

```diff
@@ -0,0 +1,194 @@
+
+    def get_children(self):
+        yield from self.subqueries
+
+
+class SearchQueryShortcut(SearchQuery):
+    def get_equivalent(self):
+        raise NotImplementedError
+
+    def get_children(self):
+        yield self.get_equivalent()
+
+#
+# Operators
+#
+
+
+class And(MultiOperandsSearchQueryOperator):
+    pass
+
+
+class Or(MultiOperandsSearchQueryOperator):
+    pass
+
+
+class Not(SearchQueryOperator):
+    def __init__(self, subquery: SearchQuery):
+        self.subquery = subquery
+
+    def apply(self, func):
+        return func(self.__class__(self.subquery.apply(func)))
+
+    def get_children(self):
+        yield self.subquery
+
+
+#
+# Basic query classes
+#
+
+
+class MatchAll(SearchQuery):
+    def apply(self, func):
+        return self.__class__()
+
+
+class Term(SearchQuery):
+    def __init__(self, term: str, boost: float = 1):
+        self.term = term
+        self.boost = boost
+
+    def apply(self, func):
+        return func(self.__class__(self.term, self.boost))
+
+
+class Prefix(SearchQuery):
+    def __init__(self, prefix: str, boost: float = 1):
+        self.prefix = prefix
+        self.boost = boost
+
+    def apply(self, func):
+        return func(self.__class__(self.prefix, self.boost))
+
+
+class Fuzzy(SearchQuery):
+    def __init__(self, term: str, max_distance: float = 3, boost: float = 1):
+        self.term = term
+        self.max_distance = max_distance
+        self.boost = boost
+
+    def apply(self, func):
+        return func(self.__class__(self.term, self.max_distance, self.boost))
+
+
+#
+# Shortcut query classes
+#
+
+
+class PlainText(SearchQueryShortcut):
+    OPERATORS = {
+        'and': And,
+        'or': Or,
+    }
+    DEFAULT_OPERATOR = 'and'
+
+    def __init__(self, query_string: str, operator: str = DEFAULT_OPERATOR,
+                 boost: float = 1):
+        self.query_string = query_string
+        self.operator = operator.lower()
+        if self.operator not in self.OPERATORS:
+            raise ValueError("`operator` must be either 'or' or 'and'.")
+        self.boost = boost
+
+    def apply(self, func):
+        return func(self.__class__(self.query_string, self.operator,
+                                   self.boost))
+
+    def get_equivalent(self):
+        return self.OPERATORS[self.operator]([
+            Term(term, boost=self.boost)
+            for term in self.query_string.split()])
+
+
+class Filter(SearchQueryShortcut):
+    def __init__(self, query: SearchQuery,
+                 include: SearchQuery = None, exclude: SearchQuery = None):
+        self.query = query
+        self.include = include
+        self.exclude = exclude
+
+    def apply(self, func):
+        return func(self.__class__(
+            self.query.apply(func),
+            self.include.apply(func), self.exclude.apply(func)))
+
+    def get_equivalent(self):
+        query = self.query
+        if self.include is not None:
+            query &= self.include
```

Comment:
This should be ``query &= Boost(self.include, 0.0)`` as the score from the "include" subquery should be ignored. We might want to do this to "exclude" as well (in case of ``Filter(query, exclude=Not(other_query)``)